CA2208 is not reported for .ctor(string? paramName, string? message) when exception derives from ArgumentException

### Analyzer

**Diagnostic ID**: [CA2208](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2208): `Instantiate argument exceptions correctly'

### Analyzer source

**NuGet Package**: [Microsoft.CodeAnalysis.NetAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.NetAnalyzers)

**Version**: 8.0.0-preview1.23121.1

### Describe the bug

**CA2208** is not reported when an incorrect parameter name is specified in `.ctor(string? paramName, string? message)` of an exception that derives from **ArgumentException**.

### Steps To Reproduce

1. Use the following .editorconfig fragment to configure **CA2208** with severity **error**:
```txt
dotnet_diagnostic.CA2208.severity = error
```
2. Compile the following code:
```c#
using System;

public sealed class InstantiateCorrectly
{
    public void Reproduce(string name)
    {
        if (name is null)
        {
            throw new ArgumentNullException("Not name", "Must not be null.");
        }
    }
}
```

### Expected behavior

The folllowing diagnostic is reported:

error CA2208: Method Reproduce passes 'Not name' as the paramName argument to a ArgumentNullException constructor. Replace this argument with one of the method's parameter names. Note that the provided parameter name should have the exact casing as declared on the method.

### Actual behavior

No diagnostic is reported.

**CA2208** is reported just fine if you change the type of the exception to **ArgumentException**:
```c#
using System;

public sealed class InstantiateCorrectly
{
    public void Reproduce(string name)
    {
        if (name is null)
        {
            throw new ArgumentException("Must not be null.", "Not name");
        }
    }
}
```

**CA2208** is also reported if you use the `.ctor(string? paramName)` of **ArgumentNullException**:
```c#
using System;

public sealed class InstantiateCorrectly
{
    public void Reproduce(string name)
    {
        if (name is null)
        {
            throw new ArgumentNullException("Not name");
        }
    }
}
```

--- a/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/InstantiateArgumentExceptionsCorrectly.cs
+++ b/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/InstantiateArgumentExceptionsCorrectly.cs
@@ -1,10 +1,8 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.
 
 using System;
-using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Globalization;
-using System.Linq;
 using Analyzer.Utilities;
 using Analyzer.Utilities.Extensions;
 using Microsoft.CodeAnalysis;
@@ -105,7 +103,7 @@ namespace Microsoft.NetCore.Analyzers.Runtime
             }
             else
             {
-                List<Diagnostic> diagnostics = new();
+                Diagnostic? diagnosticFound = null;
                 foreach (IArgumentOperation argument in creation.Arguments)
                 {
                     if (argument.Parameter?.Type.SpecialType != SpecialType.System_String)
@@ -123,21 +121,18 @@ namespace Microsoft.NetCore.Analyzers.Runtime
 
                     if (diagnostic != null)
                     {
+                        diagnosticFound = diagnostic;
                         // RuleIncorrectMessage is the highest priority rule, no need to check other rules
                         if (diagnostic.Descriptor.Equals(RuleIncorrectMessage))
                         {
-                            context.ReportDiagnostic(diagnostic);
-                            return;
+                            break;
                         }
-
-                        diagnostics.Add(diagnostic);
                     }
                 }
 
-                if (diagnostics.Count != 0)
+                if (diagnosticFound != null)
                 {
-                    // Report the last found diagnostic otherwise
-                    context.ReportDiagnostic(diagnostics.Last());
+                    context.ReportDiagnostic(diagnosticFound);
                 }
             }
         }

--- a/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/InstantiateArgumentExceptionsCorrectly.cs
+++ b/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/InstantiateArgumentExceptionsCorrectly.cs
@@ -1,8 +1,10 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.
 
 using System;
+using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Globalization;
+using System.Linq;
 using Analyzer.Utilities;
 using Analyzer.Utilities.Extensions;
 using Microsoft.CodeAnalysis;
@@ -103,7 +105,7 @@ namespace Microsoft.NetCore.Analyzers.Runtime
             }
             else
             {
-                Diagnostic? diagnostic = null;
+                List<Diagnostic> diagnostics = new();
                 foreach (IArgumentOperation argument in creation.Arguments)
                 {
                     if (argument.Parameter?.Type.SpecialType != SpecialType.System_String)
@@ -117,18 +119,25 @@ namespace Microsoft.NetCore.Analyzers.Runtime
                         continue;
                     }
 
-                    diagnostic = CheckArgument(owningSymbol, creation, argument.Parameter, value, context);
+                    Diagnostic? diagnostic = CheckArgument(owningSymbol, creation, argument.Parameter, value, context);
 
-                    // RuleIncorrectMessage is the highest priority rule, no need to check other rules
-                    if (diagnostic != null && diagnostic.Descriptor.Equals(RuleIncorrectMessage))
+                    if (diagnostic != null)
                     {
-                        break;
+                        // RuleIncorrectMessage is the highest priority rule, no need to check other rules
+                        if (diagnostic.Descriptor.Equals(RuleIncorrectMessage))
+                        {
+                            context.ReportDiagnostic(diagnostic);
+                            return;
+                        }
+
+                        diagnostics.Add(diagnostic);
                     }
                 }
 
-                if (diagnostic != null)
+                if (diagnostics.Count != 0)
                 {
-                    context.ReportDiagnostic(diagnostic);
+                    // Report the last found diagnostic otherwise
+                    context.ReportDiagnostic(diagnostics.Last());
                 }
             }
         }

--- a/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/InstantiateArgumentExceptionsCorrectlyTests.cs
+++ b/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/InstantiateArgumentExceptionsCorrectlyTests.cs
@@ -1,4 +1,4 @@
-// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.
 
 using System.Threading.Tasks;
 using Microsoft.CodeAnalysis.Testing;
@@ -1091,6 +1091,20 @@ dotnet_code_quality.CA2208.api_surface = public") }
                         }
                     }
                 }");
+        }
+
+        [Fact, WorkItem(6580, "https://github.com/dotnet/roslyn-analyzers/issues/6580")]
+        public async Task ArgumentNullException_Test()
+        {
+            await VerifyCS.VerifyAnalyzerAsync(@"
+                public class Class
+                {
+                    public void Test(string name)
+                    {
+                        if (name is null) throw new System.ArgumentNullException(""not name"", ""Must not be null."");
+                    }
+                }",
+                GetCSharpIncorrectParameterNameExpectedResult(6, 49, "Test", "not name", "paramName", "ArgumentNullException"));
         }
 
         private static DiagnosticResult GetCSharpNoArgumentsExpectedResult(int line, int column, string typeName) =>

